heat/core/statistics.py

```diff
@@ -626,77 +627,94 @@ def maximum(x1, x2, out=None):
     >>> ht.maximum(a, d)
     ValueError: operands could not be broadcast, input shapes (3, 4) (3, 4, 5)
     """
-    # perform sanitation
-    if not isinstance(x1, dndarray.DNDarray) or not isinstance(x2, dndarray.DNDarray):
-        raise TypeError(
-            "expected x1 and x2 to be a ht.DNDarray, but were {}, {} ".format(type(x1), type(x2))
-        )
-    if out is not None and not isinstance(out, dndarray.DNDarray):
-        raise TypeError("expected out to be None or an ht.DNDarray, but was {}".format(type(out)))
-
-    # apply split semantics
-    if x1.split is not None or x2.split is not None:
-        if x1.split is None:
-            x1.resplit_(x2.split)
-        if x2.split is None:
-            x2.resplit_(x1.split)
-        if x1.split != x2.split:
-            if np.prod(x1.gshape) < np.prod(x2.gshape):
-                x1.resplit_(x2.split)
-            if np.prod(x2.gshape) < np.prod(x1.gshape):
-                x2.resplit_(x1.split)
+    # TODO: move all of this to __binary_op
```

Comment:
do you want to move this logic there before or after this is merged? have you done speed comparisons between the old and new implementations?

Reply:
Hi @coquelin77 , thanks for the comments. 
I've just run a very rudimentary check on my laptop. Here's the code:
```
x = ht.array(np.random.randn(20000, 2000), dtype=ht.float32, split=0)
y = ht.array(np.random.randn(20000, 2000), dtype=ht.float32, split=0)

repeats = 10
ht.minimum(x, y) #warm-up

start = time.time()
for i in range(repeats):
    z = ht.minimum(x, y)
end = time.time()

print("Minimum took ", (end - start) / repeats, " sec on average with ", x.comm.size, " processes and split ", x.split)
```
And here are the results:
**on the main branch**
```
(py37) Claudias-MBP:heat c.comito$ mpirun --tag-output -n 2 --use-hwthread-cpus python local_test.py
[1,0]<stdout>:Minimum took  0.16146039962768555  sec on average with  2  processes and split  0
[1,1]<stdout>:Minimum took  0.1614467144012451  sec on average with  2  processes and split  0

(py37) Claudias-MBP:heat c.comito$ mpirun --tag-output -n 4 --use-hwthread-cpus python local_test.py
[1,1]<stdout>:Minimum took  0.2378169059753418  sec on average with  4  processes and split  0
[1,2]<stdout>:Minimum took  0.23781290054321289  sec on average with  4  processes and split  0
[1,0]<stdout>:Minimum took  0.23747990131378174  sec on average with  4  processes and split  0
[1,3]<stdout>:Minimum took  0.23850111961364745  sec on average with  4  processes and split  0
```
**On this branch (bug/minimum-split):**
```
(py37) Claudias-MBP:heat c.comito$ git checkout bug/minimum-split
Switched to branch 'bug/minimum-split'
Your branch is up to date with 'origin/bug/minimum-split'.
(py37) Claudias-MBP:heat c.comito$ mpirun --tag-output -n 2 --use-hwthread-cpus python local_test.py
[1,0]<stdout>:Minimum took  0.058056998252868655  sec on average with  2  processes and split  0
[1,1]<stdout>:Minimum took  0.05825142860412598  sec on average with  2  processes and split  0

(py37) Claudias-MBP:heat c.comito$ mpirun --tag-output -n 4 --use-hwthread-cpus python local_test.py
[1,0]<stdout>:Minimum took  0.05380411148071289  sec on average with  4  processes and split  0
[1,2]<stdout>:Minimum took  0.0541417121887207  sec on average with  4  processes and split  0
[1,3]<stdout>:Minimum took  0.05416550636291504  sec on average with  4  processes and split  0
[1,1]<stdout>:Minimum took  0.05364921092987061  sec on average with  4  processes and split  0
```
(yes yes I know I can format the floats :p )

Note that I haven't got 4 CPUs, those are 4 hardware threads.

I'm quite surprised it makes such a difference tbh, but anyway, there you have it.